lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -1096,14 +1096,51 @@ OpFoldResult BitsPrimOp::fold(ArrayRef<Attribute> operands) {
 LogicalResult BitsPrimOp::canonicalize(BitsPrimOp op,
                                        PatternRewriter &rewriter) {
   auto *inputOp = op.getInput().getDefiningOp();
+  if (!inputOp)
+    return failure();
   // bits(bits(x, ...), ...) -> bits(x, ...).
-  if (auto innerBits = dyn_cast_or_null<BitsPrimOp>(inputOp)) {
+  if (auto innerBits = dyn_cast<BitsPrimOp>(inputOp)) {
     auto newLo = op.getLo() + innerBits.getLo();
     auto newHi = newLo + op.getHi() - op.getLo();
     replaceOpWithNewOpAndCopyName<BitsPrimOp>(
         rewriter, op, innerBits.getInput(), newHi, newLo);
     return success();
   }
+  // bits(mux(sel, a, b), ...) -> mux(sel, bits(a, ...), bits(b, ...)).
+  if (auto mux = dyn_cast<MuxPrimOp>(inputOp)) {
+    auto bitsHigh = rewriter.create<BitsPrimOp>(op->getLoc(), mux.getHigh(),
+                                                op.getHi(), op.getLo());
+    auto bitsLow = rewriter.create<BitsPrimOp>(op->getLoc(), mux.getLow(),
+                                               op.getHi(), op.getLo());
+    replaceOpWithNewOpAndCopyName<MuxPrimOp>(rewriter, op, mux.getSel(),
+                                             bitsHigh, bitsLow);
+    return success();
+  }
+  // bits(cat(a, b), ...) -> bits(a, ...), or bits(b, ...), or cat(bits(a, ...),
```

Comment:
This is a very useful rewrite! Just FYI, it also works for `add`, `sub` and `mul` if the low bit is zero.

Also, since I am unfamiliar with this code base, I assume that when this is run, `tail` and `head` have been converted to `bits`, correct? Otherwise you might miss some optimizations.

Reply:
I think it might be a good idea to actually remove these canonicalizations that I added from `FIRRTLFolds.cpp`, since the canonicalizations that are necessary for subword initialization checking must be directly in the ExpandWhens pass, and having both creates code duplication.

Having these canonicalizations generally useful though, and should probably go in https://github.com/llvm/circt/blob/93112d40df2bd95058eae32566c13733f9fa7ea1/lib/Dialect/Comb/CombFolds.cpp#L539 instead (I think these particular ones are actually already there). I don't think it currently has the canonicalizations with `add`, `sub`, and `mul`, but it would be cool to add those!

When ExpandWhens runs, `tail` and `head` are still present, but I don't think they need to be handled by the initialization-checking canonicalizations. I believe they do get transformed into `comb.extract` ops at a later stage, and then have the full optimizing canonicalizations applied to them.